### DISCUSSION TITLE
Release v4.1

### DISCUSSION THREAD

**Discussion by tylov** [stclib/STC#47] [Announcements]
I am happy to finally announce a new release! Major changes:
- A new exciting [**cspan**](docs/cspan_api.md) single/multi-dimensional array view (with numpy-like slicing).
- Signed sizes and indices for all containers. See C++ Core Guidelines by Stroustrup/Sutter: [ES.100](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es100-dont-mix-signed-and-unsigned-arithmetic), [ES.102](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es102-use-signed-types-for-arithmetic), [ES.106](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es106-dont-try-to-avoid-negative-values-by-using-unsigned), and [ES.107](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es107-dont-use-unsigned-for-subscripts-prefer-gslindex).
- Customizable allocator [per templated container type](https://github.com/tylov/STC/discussions/44#discussioncomment-4891925).
- Updates on **cregex** with several [new unicode character classes](docs/cregex_api.md#regex-cheatsheet).
- Algorithms:
    - [crange](docs/ccommon_api.md#crange) - similar to [boost::irange](https://www.boost.org/doc/libs/release/libs/range/doc/html/range/reference/ranges/irange.html) integer range generator.
    - [c_forfilter](docs/ccommon_api.md#c_forfilter) - ranges-like filtering.
    - [csort](misc/benchmarks/various/csort_bench.c) - fast quicksort with custom inline comparison.
- Support for [uppercase flow-control](include/stc/priv/altnames.h) macro names in ccommon.h.
- Create single header container versions with python script.
- Some API changes in cregex and cstr.

<hr /><em>This discussion was created from the release <a href='https://github.com/tylov/STC/releases/tag/v4.1'>Release v4.1</a>.</em>

**Comment by murtazahakim** [comment]
Absolutely wonderful.
